### PR TITLE
Runtime changes to enable protoc integration

### PR DESCRIPTION
**Pull request by buildbreaker**
This mainly extracts out the `SchemaHandler.Context` into a separate `SchemaContext` interface to open an entry point from protoc into wire's code generator and linker.

The change here is primarily a method signature change to accept an interface rather than a direct implementaiton.

### CODE REVIEW

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/SchemaContext.kt [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * limitations under the License.
+ */
+package com.squareup.wire.schema
+
+import com.squareup.wire.WireLogger
+import okio.Path
+
+interface SchemaContext {
+  /** Location on [fileSystem] where the [SchemaHandler] is to write files, if it needs to. */
+  val outDirectory: Path
```
Comment (jhump) [inline wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/SchemaContext.kt]: Shouldn't this be configured on the concrete `FileSystemContext` and not on the interface? How would this get used by a protoc plugin?

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/SchemaContext.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire.schema
+
+import com.squareup.wire.WireLogger
+import okio.Path
+
+interface SchemaContext {
```
Comment (jhump) [inline wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/SchemaContext.kt]: There are a _lot_ of attributes on this. Do we create an implementation that needs all of these? Do some of these belong somewhere else? This feels like a complicated abstraction, and I can't really tell exactly how we'll use it. I looked in #28 and can't find any implementation of this interface there.

**Review by jhump — COMMENTED** [review]
I don't understand enough about the intended "public" API of these modules. This seems backwards-incompatible, but perhaps it is a package that is not intended for external use? Also, it's unclear how/why this abstraction is needed.

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/SchemaContext.kt (hunk 2 above)
Comment (buildbreaker) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/SchemaContext.kt]: This is a pretty complicated abstraction. I've isolated all the file system generation type things out of this interface which does simplify things a little bit.

The implementation for #28 is here: https://github.com/buildbreaker/wire/pull/28/files#diff-03e95ec28fc5596050bdb6007c807de451e9c070ea7bd526d404f7d14cc18875R65-R98

I can move it out into a new file to make it clearer in that pr as well.

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/SchemaContext.kt (hunk 1 above)
Comment (buildbreaker) [reply wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/SchemaContext.kt]: I moved this and the directory creation under that schema context implementation

File: wire-compiler/src/main/java/com/squareup/wire/schema/Target.kt [hunk 3]
```diff
@@ -400,28 +382,24 @@ data class SwiftTarget(
...
+        val filePath = "${swiftFile.name}.swift".toPath()
+        context.logArtifactHandled(
+          "${swiftFile.moduleName}.${typeName.canonicalName}", "Swift"
+        )
+        return try {
+          context.write(filePath, swiftFile.toString())
         } catch (e: IOException) {
           throw IOException(
-            "Error emitting ${swiftFile.moduleName}.${typeName.canonicalName} to $modulePath", e
+            "Error emitting ${swiftFile.moduleName}.${typeName.canonicalName} to $filePath", e
```
Comment (jhump) [inline wire-compiler/src/main/java/com/squareup/wire/schema/Target.kt]: Shouldn't this be `$outDirectory`? (The previous output, `$modulePath`, was set to `context.outDirectory`, not the file path.)

File: wire-compiler/src/main/java/com/squareup/wire/schema/Target.kt [hunk 4]
```diff
@@ -153,31 +151,25 @@ data class JavaTarget(
...
-          context.fileSystem.createDirectories(filePath.parent!!)
-          context.fileSystem.write(filePath) {
-            writeUtf8(javaFile.toString())
-          }
+        return try {
+          context.write(filePath, javaFile.toString())
         } catch (e: IOException) {
           throw IOException(
-            "Error emitting ${javaFile.packageName}.${javaFile.typeSpec.name} to $outDirectory", e
+            "Error emitting ${javaFile.packageName}.${javaFile.typeSpec.name} to ${outDirectory.toPath()}", e
```
Comment (jhump) [inline wire-compiler/src/main/java/com/squareup/wire/schema/Target.kt:170]: What will this actually log when invoked as a plugin? What will the target's `outDirectory` be set to?

File: wire-compiler/src/main/java/com/squareup/wire/schema/Target.kt [hunk 5]
```diff
@@ -155,29 +153,23 @@ data class JavaTarget(
...
         val javaFile = JavaFile.builder(javaTypeName.packageName(), typeSpec)
           .addFileComment("\$L", WireCompiler.CODE_GENERATED_BY_WIRE)
           .addFileComment("\nSource: \$L in \$L", source, location.withPathOnly())
           .build()
-        val filePath = outDirectory /
-          javaFile.packageName.replace(".", "/") /
+        val filePath = javaFile.packageName.replace(".", "/").toPath() /
           "${javaTypeName.simpleName()}.java"
 
-        context.logger.artifactHandled(
```
Comment (jhump) [inline wire-compiler/src/main/java/com/squareup/wire/schema/Target.kt:167]: Is there a reason to change this? Why not leave the `logger` field on context? When you create the context for the protoc plugin invocation, you can provide a custom implementation of `WireLogger` if you need. It seems strange to mix these methods into the file writer -- writing generated code and logging seem like two separate concerns that don't need to be conjoined into the same interface.

In the other PR, it looked like your custom implementation was just ignoring this log call. So you could leave the `logger` field and set it to `NULL_LOGGER`.

File: wire-compiler/src/main/java/com/squareup/wire/schema/WireRun.kt [hunk 6]
```diff
@@ -271,6 +273,8 @@ data class WireRun(
           module = module,
           profileLoader = schemaLoader,
         )
+        // Ensure the out directory is created.
+        fs.createDirectories(outDirectory)
```
Comment (jhump) [inline wire-compiler/src/main/java/com/squareup/wire/schema/WireRun.kt:277]: This was previously done lazily, right? So if no files were actually generated, it would not end up creating the door output directory. Is this change intentional? Is it not sufficient to lazily create directories right before creating generated files?

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/SchemaHandler.kt [hunk 7]
```diff
@@ -65,13 +63,8 @@ abstract class SchemaHandler {
...
    */
-  data class Context(
-    /** To be used by the [SchemaHandler] for reading/writing operations on disk. */
-    val fileSystem: FileSystem,
-    /** Location on [fileSystem] where the [SchemaHandler] is to write files, if it needs to. */
-    val outDirectory: Path,
-    /** Event-listener like logger with which [SchemaHandler] can notify handled artifacts. */
-    val logger: WireLogger,
+  data class Context constructor(
+    val fileWriter: FileWriter,
```
Comment (jhump) [inline wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/SchemaHandler.kt:69]: doc comment

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/SchemaHandler.kt [hunk 8]
```diff
@@ -65,13 +63,8 @@ abstract class SchemaHandler {
    * A [Context] holds the information necessary for a [SchemaHandler] to do its job. It contains
    * both helping objects such as [logger], and constraining objects such as [emittingRules].
    */
-  data class Context(
-    /** To be used by the [SchemaHandler] for reading/writing operations on disk. */
-    val fileSystem: FileSystem,
-    /** Location on [fileSystem] where the [SchemaHandler] is to write files, if it needs to. */
-    val outDirectory: Path,
-    /** Event-listener like logger with which [SchemaHandler] can notify handled artifacts. */
-    val logger: WireLogger,
```
Comment (jhump) [inline wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/SchemaHandler.kt:74]: As mentioned above, I think we want to keep this and remove the `logArtifactHandled` method from `FireWriter`.

File: wire-schema/src/jvmMain/kotlin/com/squareup/wire/schema/CoreLoader.kt [hunk 9]
```diff
@@ -44,7 +44,7 @@ actual object CoreLoader : Loader {
       }
     }
 
-    throw error("unexpected load: $path")
+    error("unexpected load: $path")
```
Comment (jhump) [inline wire-schema/src/jvmMain/kotlin/com/squareup/wire/schema/CoreLoader.kt]: Was this change intentional? Does this actually do anything w/out the `throw` keyword?

File: wire-compiler/src/main/java/com/squareup/wire/schema/WireRun.kt (hunk 6 above)
Comment (buildbreaker) [reply wire-compiler/src/main/java/com/squareup/wire/schema/WireRun.kt:277]: There's actually a test which I'm trying to keep passing which I think assumes this type of behavior. Let's keep note on this and surface this up with the block folks? I can make the change to that behavior on whatever we decide

File: wire-schema/src/jvmMain/kotlin/com/squareup/wire/schema/CoreLoader.kt (hunk 9 above)
Comment (buildbreaker) [reply wire-schema/src/jvmMain/kotlin/com/squareup/wire/schema/CoreLoader.kt]: It throws underneath the method itself:
```@kotlin.internal.InlineOnly
public inline fun error(message: Any): Nothing = throw IllegalStateException(message.toString())```

File: wire-compiler/src/main/java/com/squareup/wire/schema/Target.kt (hunk 5 above)
Comment (buildbreaker) [reply wire-compiler/src/main/java/com/squareup/wire/schema/Target.kt:167]: The `NULL_LOGGER` is for tests but we can do something similar for the protoc case

File: wire-schema/src/jvmMain/kotlin/com/squareup/wire/schema/CoreLoader.kt (hunk 9 above)
Comment (buildbreaker) [reply wire-schema/src/jvmMain/kotlin/com/squareup/wire/schema/CoreLoader.kt]: oops

Comment (buildbreaker) [reply wire-schema/src/jvmMain/kotlin/com/squareup/wire/schema/CoreLoader.kt]: image

Comment (buildbreaker) [reply wire-schema/src/jvmMain/kotlin/com/squareup/wire/schema/CoreLoader.kt]: I'll remove this change

File: wire-compiler/src/main/java/com/squareup/wire/schema/Target.kt (hunk 4 above)
Comment (buildbreaker) [reply wire-compiler/src/main/java/com/squareup/wire/schema/Target.kt:170]: This is the bit that is a bit weird -- I'm making the assumption on this stage that the `outDirectory` is what is used to create the `FileWriter` to preserve some of the error messaging that was in place prior

**Review by jhump — COMMENTED** [review]
LGTM! Hopefully it will LG to Square folks, too.
